File: AgentenSimulation_1/src/Verhandlung.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.*;
import java.util.concurrent.*;


//SIMULATION!

/*
 * Was ist das "Problem" der nachfolgenden Verhandlung?
 * - Frühe Stagnation, da die Agenten frühzeitig neue Contracte ablehnen
 * - Verhandlung ist nur für wenige Agenten geeignet, da mit Anzahl Agenten auch die Stagnationsgefahr wächst
 *
 * Aufgabe: Entwicklung und Anaylse einer effizienteren Verhandlung. Eine Verhandlung ist effizienter, wenn
 * eine frühe Stagnation vermieden wird und eine sozial-effiziente Gesamtlösung gefunden werden kann.
 *
 * Ideen:
 * - Agenten müssen auch Verschlechterungen akzeptieren bzw. zustimmen, die einzuhaltende Mindestakzeptanzrate wird vom Mediator vorgegeben
 * - Agenten schlagen alternative Kontrakte vor
 * - Agenten konstruieren gemeinsam einen Kontrakt
 * - In jeder Runde werden mehrere alternative Kontrakte vorgeschlagen
 * - Alternative Konstruktionsmechanismen für Kontrakte
 * - Ausgleichszahlungen der Agenten (nur möglich, wenn beide Agenten eine monetaere Zielfunktion haben
 *
 */


public class Verhandlung {
    public static final long maxMemBytes = Runtime.getRuntime().maxMemory();
    public static final int ContractObjectMemSizeBytes = 985;

    //Hyperparameter
    private static final int generationsSize = 500_000;
    private static final int maxGenerations = 3400;

    private static final double infillRate = 0.05;
    private static final double mutationRate = 0.5; //increases to 1 during runtime

    private static final double minAcceptacneRate = 0.02;
    private static final double maxAcceptacneRate = 0.7;
    private static final double acceptanceRateGrowth = maxAcceptacneRate - minAcceptacneRate;
    private static final double accepanceRateOffset = 0.05;

    public static void main(String[] args) {
        final long completeRuntimeStart = System.nanoTime();
        Contract[] generation;  //ein Contract Object ~971 bytes bei int[200]
        Agent agA, agB;
        Mediator med;
        int currentAcceptanceAmount = (int) (generationsSize * 0.77);//0.77
        //int currentInfill = (int) (generationsSize * infillRate);
        int currentInfill = 0;
        int mutationAmount = (int) (generationsSize * mutationRate);
        SplittableRandom rand = new SplittableRandom();

        int avpro = Runtime.getRuntime().availableProcessors();
        try (ExecutorService executor = Executors.newFixedThreadPool(avpro)) {
            agA = new SupplierAgent(new File("../data/daten3ASupplier_200.txt"));
            //agB = new SupplierAgent(new File("../data/daten3ASupplier_200.txt"));
            //agA = new CustomerAgent(new File("../data/daten4BCustomer_200_5.txt"));
            agB = new CustomerAgent(new File("../data/daten4BCustomer_200_5.txt"));
            med = new Mediator(agA.getContractSize(), agB.getContractSize(), generationsSize);

            //Verhandlung initialisieren
            generation = med.initContract();
            Set<Contract> knownBadContracts = ConcurrentHashMap.newKeySet();

            for (int currentGeneration = 0; currentGeneration < maxGenerations; currentGeneration++) {
                //Step1: Update Relevant Variables for usage of simulated annealing
                final double generationProgress = (double) currentGeneration / maxGenerations;
                currentAcceptanceAmount = Math.min(
                        (int) ((double) generationsSize * maxAcceptacneRate),
                        Math.max(
                                (int) (generationsSize * minAcceptacneRate),
                                (int) (generationsSize * (
                                        (1 - (((double) currentGeneration / maxGenerations)))
                                                * (acceptanceRateGrowth + accepanceRateOffset)
                                                + (minAcceptacneRate - accepanceRateOffset))
                                )
                        ));
                mutationAmount = Math.min((int) (generationsSize * mutationRate), (int) (generationsSize * (((double) currentGeneration / maxGenerations))));
                //currentInfill = Math.min((int) (generationsSize * 0.7), (int) ((generationsSize * (((double) currentGeneration / maxGenerations)))*0.3));

                if ((long)knownBadContracts.size() * ContractObjectMemSizeBytes > (double)maxMemBytes*0.3) {  //Max 30% of Heap Space
                    knownBadContracts.clear();  //Clearing Memmory
                    System.out.println("Known Bad Cleared");
                }


                /*Step 2: Let Agents vote on the generation
                    Return which Contracts are wanted by the Agents
                    Primarily using intersect to move forward -> Steady State GA
                    Filter out known bad contracts -> not shown again to agent -> more/ better contracts can be shown
                 */
                long startTime = System.nanoTime();

                int finalCurrentAcceptanceAmount = currentAcceptanceAmount;
                Contract[] finalGeneration = generation;
                CompletableFuture<boolean[]> voteAFuture = CompletableFuture.supplyAsync(() -> agA.voteLoop(finalGeneration, finalCurrentAcceptanceAmount), executor);
                CompletableFuture<boolean[]> voteBFuture = CompletableFuture.supplyAsync(() -> agB.voteLoop(finalGeneration, finalCurrentAcceptanceAmount), executor);

                CompletableFuture.allOf(voteAFuture, voteBFuture).join();
                boolean[] voteA = voteAFuture.get();
                boolean[] voteB = voteBFuture.get();

                long voteTime = System.nanoTime() - startTime;
                ArrayList<Contract> intersect = new ArrayList<>();

                int voteACount = 0;
                for (boolean vote : voteA) {
                    voteACount+=vote?1:0;
                }
                int voteBCount = 0;
                for (boolean vote : voteB) {
                    voteBCount+=vote?1:0;
                }

                System.out.println(voteACount + " " + voteBCount);

                for (int i = 0; i < generationsSize; i++) {
                    if (voteA[i] && voteB[i]) {
                        intersect.add(generation[i]);
                    } else if (!voteA[i] && !voteB[i]) {
                        knownBadContracts.add(generation[i]);   //Just store the ones no agents want
                    }
                }
                int intersectSize = intersect.size();

                if (intersect.isEmpty()) {
                    //TODO
                    throw new UnsupportedOperationException("Feature incomplete. Contact assistance.");
                }
                Contract printIntersect = intersect.getFirst();

                //Get best 10% of intersect -> can be used for a higher selection probability for these Contracts
                ArrayList<Contract> bestIntersect = new ArrayList<>();
                CompletableFuture<boolean[]> voteAFutureBest = CompletableFuture.supplyAsync(() -> agA.voteLoop(intersect.toArray(new Contract[intersectSize]), (int) (intersectSize*0.1)), executor);
                CompletableFuture<boolean[]> voteBFutureBest = CompletableFuture.supplyAsync(() -> agB.voteLoop(intersect.toArray(new Contract[intersectSize]), (int) (intersectSize*0.1)), executor);
                CompletableFuture.allOf(voteAFutureBest, voteBFutureBest).join();

                boolean[] voteABest = voteAFutureBest.get();
                boolean[] voteBBest = voteBFutureBest.get();
                for (int i = 0; i < intersectSize; i++) {
                    if (voteABest[i] && voteBBest[i]) {
                        bestIntersect.add(generation[i]);       //Steady State GA
                        bestIntersect.add(generation[i]);
                    } else if (voteABest[i] || voteBBest[i]) {
                        bestIntersect.add(generation[i]);
                    }
                }

                //List of all the valid Contracts that are used for Mating (with right Probabilities)
                List<Contract> proportionalCrossOverSelektion = new ArrayList<>(intersect);
                proportionalCrossOverSelektion.addAll(bestIntersect);
                //proportionalCrossOverSelektion.addAll(bestIntersect);
                Collections.shuffle(proportionalCrossOverSelektion);    //maybe not relevant since always two random Contracts are picked for mating

                /*
                Step 3: Create new Generation
                    - CrossOver
                    - Mutation
                 */

                Set<Contract> newGenerationHashSet = ConcurrentHashMap.newKeySet();
                if (generationProgress > 0.8) {
                    newGenerationHashSet.addAll(intersect);
                }

                List<CompletableFuture<Void>> futures = new LinkedList<>();
                int whileCount = 0;
                while (newGenerationHashSet.size() < generationsSize && whileCount < generationsSize * 3) {
                    int innerWhileCount = 0;
                    int newGenerationHashSetStartsize = newGenerationHashSet.size();
                    while (newGenerationHashSet.size() < generationsSize && innerWhileCount < (generationsSize-newGenerationHashSetStartsize)) {
                        futures.add(
                                CompletableFuture.supplyAsync(() -> {
                                    Contract parent1 = proportionalCrossOverSelektion.get(rand.nextInt(proportionalCrossOverSelektion.size()));
                                    Contract parent2 = proportionalCrossOverSelektion.get(rand.nextInt(proportionalCrossOverSelektion.size()));
                                    Contract[] childs = parent1.crossover(parent2);

                                    if(!knownBadContracts.contains(childs[0])){
                                        newGenerationHashSet.add(childs[0]);
                                    }
                                    if(!knownBadContracts.contains(childs[1])){
                                        newGenerationHashSet.add(childs[1]);
                                    }

                                    return null;
                                }, executor)
                        );
                        whileCount++;
                        innerWhileCount++;
                    }
                    CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();
                }

                //if not enough contract after (generationsSize * 3) try just fill with random - just for the sake of runtime
                while (newGenerationHashSet.size() < generationsSize) {
                    newGenerationHashSet.addAll(Arrays.stream(med.getRandomContracts(generationsSize - newGenerationHashSet.size())).toList());
                }

                //Overwrite old generation Array -> mutation will happen afterward to reduce memory footprint
                generation = newGenerationHashSet.toArray(Contract[]::new);

                // Mutate
                for (int i = 0; i < mutationAmount; i++) {
                    int contractIndexToMutate = rand.nextInt(generationsSize);
                    generation[contractIndexToMutate].swapMutateRand();
                }


                System.out.printf("%4d: Best A: %5d \t Best B: %5d \t AccAmount: %4d \t Intersect: %4d \t Contract: %5d %5d",
                        currentGeneration,
                        agA.getRound_best().getCost(), agB.getRound_best().getCost(),
                        currentAcceptanceAmount, intersectSize,
                        agA.printUtility(printIntersect).getCost(), agB.printUtility(printIntersect).getCost());

                long mediatorTime = System.nanoTime() - startTime - voteTime;
                System.out.println("   Time:" + voteTime + "  " + mediatorTime);
                System.out.flush();
            }

            System.out.println("----------");
            System.out.println("Changing to Terminal Phase");
            System.out.println("----------");

            boolean[] voteA = agA.voteLoop(generation, currentAcceptanceAmount);
            boolean[] voteB = agB.voteLoop(generation, currentAcceptanceAmount);
            ArrayList<Contract> intersect = new ArrayList<>();

            for (int i = 0; i < generationsSize; i++) {
                if (voteA[i] && voteB[i]) {
                    intersect.add(generation[i]);
                }
            }

            Contract[] temp = new Contract[intersect.size()];
            generation = intersect.toArray(temp);

            while (generation.length > 1) {
                int remove;
                if (generation.length % 2 == 0) {
                    remove = agA.voteEnd(generation);
                } else {
                    remove = agB.voteEnd(generation);
                }

                Contract[] newGeneration = new Contract[generation.length - 1];
                System.arraycopy(generation, 0, newGeneration, 0, remove);
                System.arraycopy(generation, remove + 1,
                        newGeneration, remove,
                        generation.length - remove - 1);
                generation = newGeneration;
            }

            long CompleteRuntime = System.nanoTime() - completeRuntimeStart;

            System.out.print("""
                    ----------
                    Config
                    ----------
                    """);
            System.out.printf("""
                            Config:
                            MaxGenerations:    %d \t GenerationSize:       %d
                            InfillRate:        %.3f \t MutationRate:         %.3f
                            MinAcceptacneRate: %.3f \t AcceptanceRateGrowth: %.3f
                                                        
                            """,
                    maxGenerations, generationsSize,
                    infillRate, mutationRate,
                    minAcceptacneRate, acceptanceRateGrowth
            );
            System.out.print("""
                    ----------
                    Result
                    ----------
                    """);
            System.out.printf("""
                                                
                            Best Cost A: %d
                            Best Cost B: %d
                            Sum: %d

                            Picked Contract:
                            A: %d \t B: %d
                            Sum: %d
                            """,
                    agA.getGlobal_best().getCost(),
                    agB.getGlobal_best().getCost(),
                    agA.getGlobal_best().getCost() + agB.getGlobal_best().getCost(),
                    agA.printUtility(generation[0]).getCost(),
                    agB.printUtility(generation[0]).getCost(),
                    agA.printUtility(generation[0]).getCost() + agB.printUtility(generation[0]).getCost()
            );


            // TimeUnit
            long CompleteRuntimeMinutes = TimeUnit.MINUTES.convert(CompleteRuntime, TimeUnit.NANOSECONDS);
            System.out.println("Complete Runtime: " + CompleteRuntimeMinutes + " min");

        } catch (FileNotFoundException | InterruptedException | ExecutionException e) {
            System.out.println(e.getMessage());
        }
    }

    public static void ausgabe(Agent a1, Agent a2, int i, Contract contract) {
        System.out.print(i + " -> ");
        a1.printUtility(contract);
        System.out.print("  ");
        a2.printUtility(contract);
        System.out.println();
    }

}

File: AgentenSimulation_Rust/src/verhandlung.rs
mod agent;
mod contract;
mod customer_agent;
mod mediator;
mod supplier_agent;

use std::collections::HashSet;
use std::fs::File;

use crate::agent::Agent;
use crate::contract::Contract;
use crate::customer_agent::CustomerAgent;
use crate::mediator::Mediator;
use crate::supplier_agent::SupplierAgent;
use clap::Parser;
use rand::Rng;
use rayon::prelude::*;

static AGA_DEF:&str = include_str!("../../data/daten3ASupplier_200.txt");
static AGB_DEF:&str = include_str!("../../data/daten4BCustomer_200_5.txt");

#[derive(Parser)]
struct Cli {
    #[arg(long = "gen-size", default_value_t = 1000)]
    generation_size: usize,
    #[arg(long = "max-gen", default_value_t = 3000)]
    max_generations: usize,
    #[arg(long = "infill-rate", default_value_t = 0.05)]
    infill_rate: f64,
    #[arg(long = "mutation-rate", default_value_t = 0.5)]
    mutation_rate: f64,
    #[arg(long = "min-acc-rate", default_value_t = 0.15)]
    min_acceptance_rate: f64,
    #[arg(long = "max-acc-rate", default_value_t = 0.45)]
    max_acceptance_rate: f64,
}

struct Config {
    generation_size: usize,
    max_generations: usize,
    infill_rate: f64,
    mutation_rate: f64,
    min_acceptance_rate: f64,
    max_acceptance_rate: f64,
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args = Cli::parse();

    let config = Config {
        generation_size: args.generation_size,
        max_generations: args.max_generations,
        infill_rate: args.infill_rate,
        mutation_rate: args.mutation_rate,
        min_acceptance_rate: args.min_acceptance_rate,
        max_acceptance_rate: args.max_acceptance_rate,
    };
    let mut ag_a = SupplierAgent::new_from_str(&AGA_DEF)?;
    let mut ag_b = CustomerAgent::new_from_str(&AGB_DEF)?;
    
    // let mut ag_a = SupplierAgent::new_from_file(File::open("../data/daten3ASupplier_200.txt")?)?;
    // let mut ag_b = CustomerAgent::new_from_file(File::open("../data/daten4BCustomer_200_5.txt")?)?;
    // let mut ag_a = SupplierAgent::new(File::open("../data/datenSupplier_5.txt")?)?;
    // let mut ag_b = CustomerAgent::new(File::open("../data/datenCustomer_5_3.txt")?)?;
    let med = Mediator::new(ag_a.get_contract_size(), ag_b.get_contract_size())?;

    let mut generation = med.get_random_contracts(config.generation_size);

    for n_generation in 0..config.max_generations {
        let acceptance_amount = calculate_acceptance_amount(n_generation, &config);
        let mutation_amount = calculate_mutation_amount(n_generation, &config);

        let vote_a = ag_a.vote_many(&generation, acceptance_amount);

        let vote_b = ag_b.vote_many(&generation, acceptance_amount);

        let mut intersect: Vec<Contract> = Vec::new();

        for i in 0..vote_a.len() {
            if vote_a[i] && vote_b[i] {
                intersect.push(generation[i].clone())
            }
        }

        let mut new_generation_set: HashSet<Contract> = HashSet::new();
        let mut loop_count = 0usize;
        let mut rng = rand::thread_rng();

        while new_generation_set.len() < config.generation_size
            && loop_count < config.generation_size
        {
            let missing = (0..config.generation_size - new_generation_set.len());

            let mut parents: Vec<(&Contract, &Contract)> =
                Vec::with_capacity(config.generation_size - new_generation_set.len());

            for _ in 0..config.generation_size - new_generation_set.len() {
                let parent1 = &intersect[rng.gen_range(0..intersect.len())];
                let parent2 = &intersect[rng.gen_range(0..intersect.len())];
                parents.push((parent1, parent2))
            }

            let new_children: Vec<Contract> = missing
                .into_par_iter()
                .map(|i| {
                    let (parent1, parent2) = parents[i];
                    parent1.ordered_crossover(parent2)
                })
                .collect();

            new_generation_set.extend(new_children);

            loop_count += 1;
        }

        while new_generation_set.len() < config.generation_size {
            new_generation_set.extend(
                med.get_random_contracts(config.generation_size - new_generation_set.len()),
            );
        }

        let mut new_generation: Vec<Contract> = new_generation_set.into_iter().collect();

        for _ in 0..mutation_amount {
            let i_mutate: usize = rng.gen_range(0..new_generation.len());
            new_generation[i_mutate].mutate();
        }

        generation = new_generation;

        println!(
            "{:>4}: Best A: {:>5} \t Best B: {:>5} \t AccAmount: {:>4} \t Intersect: {:>4} \t",
            n_generation,
            ag_a.get_round_best().cost,
            ag_b.get_round_best().cost,
            acceptance_amount,
            intersect.len()
        );
    }

    println!("----------\nChanging to Terminal Phase\n----------");
    
    
    let mut generation_set :HashSet<Contract>= HashSet::from_iter(generation.into_iter());

    
    while generation_set.len() > 1 {
        if generation_set.len() % 2 == 0 {
            generation_set.remove(&ag_a.vote_end(&generation_set));
        } else {
            generation_set.remove(&ag_b.vote_end(&generation_set));
        }
        print!("\x1b[2K \x1b[1G Remaining Contracts: {}", generation_set.len())
    }
    
    let result = generation_set.iter().next().unwrap();
    
    
    println!();
    println!(
        r#"----------
Config
----------

Config:
MaxGenerations:    {:5 }    GenerationSize:    {}
InfillRate:        {:.3}    MutationRate:      {:.3}
MinAcceptanceRate: {:.3}    MaxAcceptanceRate: {:.3}    AcceptanceRateGrowth: {:.3}

----------
Result
----------

Best Cost A: {}
Best Cost B: {}
Sum: {}

Picked Contract:
A: {}    B: {}
Sum: {}"#,
        config.max_generations,
        config.generation_size,
        config.infill_rate,
        config.mutation_rate,
        config.min_acceptance_rate,
        config.max_acceptance_rate,
        config.max_acceptance_rate - config.min_acceptance_rate,
        ag_a.get_global_best().cost,
        ag_b.get_global_best().cost,
        ag_a.get_global_best().cost + ag_b.get_global_best().cost,
        ag_a._eval(result),
        ag_b._eval(result),
        ag_a._eval(result) + ag_b._eval(result)
    );
    Ok(())
}

fn calculate_acceptance_amount(current_generation: usize, config: &Config) -> usize {
    let current_acceptance_amount = std::cmp::min(
        (config.generation_size as f64 * config.max_acceptance_rate) as usize,
        std::cmp::max(
            (config.generation_size as f64 * config.min_acceptance_rate) as usize,
            (config.generation_size as f64
                * (1.0 - (current_generation as f64 / config.max_generations as f64))
                * config.max_acceptance_rate
                - config.min_acceptance_rate
                + config.min_acceptance_rate) as usize,
        ),
    );

    current_acceptance_amount
}

fn calculate_mutation_amount(current_generation: usize, config: &Config) -> usize {
    let mutation_amount = std::cmp::min(
        (config.generation_size as f64 * config.mutation_rate) as usize,
        (config.generation_size as f64
            * (current_generation as f64 / config.max_generations as f64)) as usize,
    );

    mutation_amount
}
